# Ejercicio 2
Para el ejercicio 2, se han seleccionado tres fuentes, cada una relacionada con la temática de los premios Nobel de Química:
- [Fuente 1](https://en.wikipedia.org/wiki/List_of_Nobel_laureates_in_Chemistry): Tabla con los ganadores de los premios Nobel en Química, clasificados por año, nombre y país de procedencia
- [Fuente 2](https://en.wikipedia.org/wiki/List_of_nominees_for_the_Nobel_Prize_in_Chemistry): Nominados al premio Nobel de Química desde 1901 a 1970. El Comité Noruego del Nobel revela la lista de nominados hasta 50 años después de la fecha de nominación. Por esta razón, hasta la fecha, se cuenta con la lista oficial de nominados hasta 1973. Sin embargo, en la fuente que se empleará, solo se tiene información hasta 1970.
- [Fuente 3](https://en.wikipedia.org/wiki/List_of_Nobel_laureates_by_university_affiliation): Tabla con todos los ganadores de los premios Nobel, clasificados acorde a su afiliación a determinado instituto o universidad. Se deberá limpiar la tabla para filtrar solo los ganadores correspondientes al premio Nobel de Química.

## a) Fuente 1

In [23]:
import pandas as pd

LaureatesLink="https://en.wikipedia.org/wiki/List_of_Nobel_laureates_in_Chemistry"
Laureates=pd.read_html(LaureatesLink,header=0,flavor='bs4')
type(Laureates)

list

In [24]:
len(Laureates)

8

In [25]:
Laureates=pd.read_html(LaureatesLink,header=0,flavor='bs4')
Laureates_DF=Laureates[0]
Laureates_DF

Year Image                               Laureate[A]      Country[B]  \
0    1901   NaN  Jacobus Henricus van 't Hoff (1852–1911)     Netherlands   
1    1902   NaN          Hermann Emil Fischer (1852–1919)         Germany   
2    1903   NaN       Svante August Arrhenius (1859–1927)          Sweden   
3    1904   NaN            Sir William Ramsay (1852–1916)  United Kingdom   
4    1905   NaN              Adolf von Baeyer (1835–1917)         Germany   
..    ...   ...                                       ...             ...   
197  2022   NaN                   Morten Meldal (b. 1954)         Denmark   
198  2022   NaN              K. Barry Sharpless (b. 1941)   United States   
199  2023   NaN               Moungi G. Bawendi (b. 1961)   United States   
200  2023   NaN                   Louis E. Brus (b. 1943)   United States   
201  2023   NaN                   Alexey Ekimov (b. 1945)          Russia   

                                          Rationale[C]    Ref  
0    "[for his] discovery of the laws of chemical d...   [18]  
1       "[for] his work on sugar and purine syntheses"   [19]  
2      "[for] his electrolytic theory of dissociation"   [20]  
3    "[for his] discovery of the inert gaseous elem...   [21]  
4    "[for] the advancement of organic chemistry an...   [22]  
..                                                 ...    ...  
197  "for the development of click chemistry and bi...  [126]  
198  "for the development of click chemistry and bi...  [126]  
199  "for the discovery and synthesis of quantum dots"  [127]  
200  "for the discovery and synthesis of quantum dots"  [127]  
201  "for the discovery and synthesis of quantum dots"  [127]  

[202 rows x 6 columns]

In [153]:
Laureates=pd.read_html(LaureatesLink,header=0,flavor='bs4')
Laureates_DF=Laureates[0]

#Eliminando las columnas "Image" y "Ref"
Laureates_DF.drop(columns=["Image","Ref"], inplace=True)

# Renombrando las columnas
Laureates_DF.rename(columns={"Year":"Year_of_award","Laureate[A]":"Laureate","Country[B]":"Country","Rationale[C]":"Rationale"},inplace=True)

#Creando dos nuevas columnas (Birth_year y Death_year a partir de datos preexistentes de la columna Laureate)
Laureates_DF_2=Laureates_DF["Laureate"].str.split("(", expand=True) 
Laureates_DF["Laureate"]=Laureates_DF_2[0]
Laureates_DF.Laureate.str.strip()

##Eliminando "b." y ")"
Laureates_DF_2[1]=Laureates_DF_2[1].str.replace(pat=r'(\Ab.)',repl='',regex=True)
Laureates_DF_2[1]=Laureates_DF_2[1].str.replace(pat=')',repl='')
Laureates_DF_2=Laureates_DF_2[1].str.split("–", expand=True)

#Columnas resultantes luego de limpiar
Laureates_DF_2

0     1
0     1852  1911
1     1852  1919
2     1859  1927
3     1852  1916
4     1835  1917
..     ...   ...
197   1954  None
198   1941  None
199   1961  None
200   1943  None
201   1945  None

[202 rows x 2 columns]

In [154]:
#Reemplazando valores "None" en la segunda columna con "Alive"
Laureates_DF_2[1]=Laureates_DF_2[1].replace({None:'Alive'})

In [155]:
#Dataframe resultante
Laureates_DF["Birth_year"]=Laureates_DF_2[0]
Laureates_DF["Death_year"]=Laureates_DF_2[1]
Laureates_DF

Year_of_award                       Laureate         Country  \
0             1901  Jacobus Henricus van 't Hoff      Netherlands   
1             1902          Hermann Emil Fischer          Germany   
2             1903       Svante August Arrhenius           Sweden   
3             1904            Sir William Ramsay   United Kingdom   
4             1905              Adolf von Baeyer          Germany   
..             ...                            ...             ...   
197           2022                 Morten Meldal          Denmark   
198           2022            K. Barry Sharpless    United States   
199           2023             Moungi G. Bawendi    United States   
200           2023                 Louis E. Brus    United States   
201           2023                 Alexey Ekimov           Russia   

                                             Rationale Birth_year Death_year  
0    "[for his] discovery of the laws of chemical d...       1852       1911  
1       "[for] his work on sugar and purine syntheses"       1852       1919  
2      "[for] his electrolytic theory of dissociation"       1859       1927  
3    "[for his] discovery of the inert gaseous elem...       1852       1916  
4    "[for] the advancement of organic chemistry an...       1835       1917  
..                                                 ...        ...        ...  
197  "for the development of click chemistry and bi...       1954      Alive  
198  "for the development of click chemistry and bi...       1941      Alive  
199  "for the discovery and synthesis of quantum dots"       1961      Alive  
200  "for the discovery and synthesis of quantum dots"       1943      Alive  
201  "for the discovery and synthesis of quantum dots"       1945      Alive  

[202 rows x 6 columns]

In [156]:
#Conviertiendo los datos de las columnas Birth_year y Death_year en números enteros
#Laureates_DF.Birth_year=pd.to_numeric(Laureates_DF.Birth_year,errors='ignore')
#Laureates_DF.Death_year=pd.to_numeric(Laureates_DF.Death_year,errors='ignore')

#Laureates_DF[Laureates_DF.columns[4:6]]=Laureates_DF.iloc[:,4:6].apply(lambda x: x.astype('Int64',errors="ignore"))
#Laureates_DF.info()

In [157]:
#DF_final
Laureates_DF

Year_of_award                       Laureate         Country  \
0             1901  Jacobus Henricus van 't Hoff      Netherlands   
1             1902          Hermann Emil Fischer          Germany   
2             1903       Svante August Arrhenius           Sweden   
3             1904            Sir William Ramsay   United Kingdom   
4             1905              Adolf von Baeyer          Germany   
..             ...                            ...             ...   
197           2022                 Morten Meldal          Denmark   
198           2022            K. Barry Sharpless    United States   
199           2023             Moungi G. Bawendi    United States   
200           2023                 Louis E. Brus    United States   
201           2023                 Alexey Ekimov           Russia   

                                             Rationale Birth_year Death_year  
0    "[for his] discovery of the laws of chemical d...       1852       1911  
1       "[for] his work on sugar and purine syntheses"       1852       1919  
2      "[for] his electrolytic theory of dissociation"       1859       1927  
3    "[for his] discovery of the inert gaseous elem...       1852       1916  
4    "[for] the advancement of organic chemistry an...       1835       1917  
..                                                 ...        ...        ...  
197  "for the development of click chemistry and bi...       1954      Alive  
198  "for the development of click chemistry and bi...       1941      Alive  
199  "for the discovery and synthesis of quantum dots"       1961      Alive  
200  "for the discovery and synthesis of quantum dots"       1943      Alive  
201  "for the discovery and synthesis of quantum dots"       1945      Alive  

[202 rows x 6 columns]

## b) Fuente 2

In [158]:
import pandas as pd

NomineesLink="https://en.wikipedia.org/wiki/List_of_nominees_for_the_Nobel_Prize_in_Chemistry"
Nominees=pd.read_html(NomineesLink,header=0,flavor='bs4')
type(Nominees)

list

In [114]:
range(len(Nominees))

range(0, 15)

In [115]:
#El primer elemento de la lista es un DF que contiene adjuntos verticalmente datos de los nominados de 1901 a 1909
Nominees[0]

Picture                          Name  \
0    1901.0                          1901   
1       NaN  Jacobus Henricus van 't Hoff   
2       NaN                  Emil Fischer   
3       NaN              Svante Arrhenius   
4       NaN                 Henri Moissan   
5       NaN            Zdenko Hans Skraup   
6       NaN           Marcellin Berthelot   
7       NaN          William Jackson Pope   
8    1902.0                          1902   
9       NaN                William Ramsay   
10      NaN              Adolf von Baeyer   
11      NaN     Theodore William Richards   
12      NaN               Rudolf Knietsch   
13      NaN           Johannes Wislicenus   
14      NaN          Oliver Wolcott Gibbs   
15      NaN        Edward Williams Morley   
16      NaN                Armand Gautier   
17   1903.0                          1903   
18      NaN         Harmon Northrop Morse   
19   1904.0                          1904   
20      NaN               Wilhelm Ostwald   
21   1905.0                          1905   
22      NaN                Eduard Buchner   
23      NaN              Dmitri Mendeleev   
24      NaN             Giacomo Ciamician   
25      NaN      Henry Louis Le Chatelier   
26   1906.0                          1906   
27      NaN                Walther Nernst   
28      NaN               Victor Grignard   
29   1907.0                          1907   
30      NaN                 Paul Sabatier   
31      NaN             Ernest Rutherford   
32      NaN                  Otto Wallach   
33      NaN                 Alfred Werner   
34      NaN         Fredrik Adolf Kjellin   
35      NaN          Stanislao Cannizzaro   
36      NaN         Sophus Mads Jørgensen   
37      NaN               William Crookes   
38      NaN                 Heinrich Caro   
39      NaN                  Adolph Frank   
40      NaN                   Samuel Eyde   
41      NaN            Kristian Birkeland   
42      NaN               Theodor Curtius   
43      NaN                  Albin Haller   
44   1908.0                          1908   
45      NaN           Richard Willstätter   
46      NaN               Frederick Soddy   
47      NaN               Albrecht Kossel   
48      NaN               Gustaf de Laval   
49      NaN                   Georg Lunge   
50      NaN        Arthur Rudolf Hantzsch   
51   1909.0                          1909   
52      NaN     Edmund Oscar von Lippmann   
53      NaN         Hans Heinrich Landolt   
54      NaN                Otto Schönherr   
55      NaN                   Karl Engler   
56      NaN       Jean-Baptiste Senderens   
57      NaN              Gabriel Bertrand   

                                                 Born  \
0                                                1901   
1              August 30, 1852 Rotterdam, Netherlands   
2                 October 9, 1852 Euskirchen, Germany   
3                   February 19, 1859 Uppsala, Sweden   
4                    September 28, 1852 Paris, France   
5                March 3, 1850 Prague, Czech Republic   
6                      October 25, 1827 Paris, France   
7                      March 31, 1870 London, England   
8                                                1902   
9                   October 2, 1852 Glasgow, Scotland   
10                   October 31, 1835 Berlin, Germany   
11  January 31, 1868 Germantown, Philadelphia, Uni...   
12                    December 13, 1854 Opole, Poland   
13                    June 24, 1835 Querfurt, Germany   
14     February 21, 1822 New York City, United States   
15  January 29, 1838 Newark, New Jersey, United St...   
16                September 23, 1837 Narbonne, France   
17                                               1903   
18  October 15, 1848 Cambridge, Vermont, United St...   
19                                               1904   
20                     September 2, 1853 Riga, Latvia   
21                                               1905   
22                       May 20, 1860

### b.2) Filtrado y concatenación de la data 

In [117]:
# Con la función cotas, se busca encontrar el intervalo de filas donde se encuentran los nominados por cada año
# Con esto, acotar los DFs asociados a cada año, en función a dicho intervalo de filas
# x: índice de la lista Nominees, y: año de nominación
# Por ejemplo, los que fueron nominados en 1901, se encuentran en las filas 1:7 del elemento 1 de la lista Nominees

def cotas(x,y):
    try:
        cota_inf=Nominees[x].index.get_loc(Nominees[x][Nominees[x]['Picture'] == y].index[0])
        cota_sup= Nominees[x].index.get_loc(Nominees[x][Nominees[x]['Picture'] == y+1].index[0])
        return range(cota_inf+1,cota_sup)
    except:
        cota_inf=Nominees[x].index.get_loc(Nominees[x][Nominees[x]['Picture'] == y].index[0])
        cota_sup= Nominees[x].shape[0]
        return range(cota_inf+1,cota_sup)
    

list(cotas(0,1901))

[1, 2, 3, 4, 5, 6, 7]

In [6]:
#Se tiene información de los nominados desde 1901 hasta 1970 en forma de 8 DFS:

#1901–1909: Nominees[0]
#1910–1919: Nominees[1]
#1920–1929: Nominees[2]
#1930–1939: Nominees[3]
#1940–1949: Nominees[4]
#1950–1959: Nominees[5]
#1960–1969: Nominees[6]
#1970: Nominees[7]

#DFs de 1901-1909
NomineesDFs_1901_1909=[Nominees[0].iloc[cotas(0,i),:] for i in list(range(1901,1910))]

# DFs de 1910 a 1969
    #Forma larga:
NomineesDFs_1910_1919=[Nominees[1].iloc[cotas(1,i),:] for i in list(range(1910,1920))]
NomineesDFs_1920_1929=[Nominees[2].iloc[cotas(2,i),:] for i in list(range(1920,1930))]
NomineesDFs_1930_1939=[Nominees[3].iloc[cotas(3,i),:] for i in list(range(1930,1940))]
NomineesDFs_1940_1949=[Nominees[4].iloc[cotas(4,i),:] for i in list(range(1940,1950))]
NomineesDFs_1950_1959=[Nominees[5].iloc[cotas(5,i),:] for i in list(range(1950,1960))]
NomineesDFs_1960_1969=[Nominees[6].iloc[cotas(6,i),:] for i in list(range(1960,1970))]
    #Forma simplificada: 
NomineesDFs_1910_1969=[Nominees[j].iloc[cotas(j,i),:] for j in list(range(1,7)) for i in list(range(1900+j*10,1910+j*10))]


#DF de 1970 
#Los datos Nominees[7].Picture no se encontraban como números reales (float)
#Por esto, se tuvo que realizar la conversión respectiva
Nominees[7].Picture=pd.to_numeric(Nominees[7].Picture,errors='coerce')
NomineesDF_1970=Nominees[7].iloc[cotas(7,1970),:]


#INTENTOS FALLIDOS:
# DFs de 1910 a 1969
#z
#NomineesDFs_1910_1969=[Nominees[x[i]].iloc[cotas(x[i],z[i]),:] for i in list(range(0,6))]
#c=[cotas(i,j) for i in list(range(0,6))]
#c=[cotas[(a,b)] for b in list(range(1910+ (a-1)*10, 1910+ a*10)) if a in x]

#A= [Nominees[X].iloc[cotas(X,i),:] for i in list(range(1960,1970))]
#x=range(1,7)
#z=list(range(1910+ (w-1)*10, 1910+ w*10) for w in x)
#a=[(j,i) for j in list(range(1,7)) for i in list(range(1900+j*10,1910+j*10))]

In [118]:
#Concatenando los DFs de cada lista
Nominees=pd.read_html(NomineesLink,header=0,flavor='bs4')
NomineesDFs_1901_1909=[Nominees[0].iloc[cotas(0,i),:] for i in list(range(1901,1910))]
NomineesDFs_1901_1909=pd.concat(NomineesDFs_1901_1909[i] for i in range(len(NomineesDFs_1901_1909))) 

NomineesDFs_1910_1969=[Nominees[j].iloc[cotas(j,i),:] for j in list(range(1,7)) for i in list(range(1900+j*10,1910+j*10))]
NomineesDFs_1910_1969=pd.concat(NomineesDFs_1910_1969[i] for i in range(len(NomineesDFs_1910_1969))) 

Nominees[7].Picture=pd.to_numeric(Nominees[7].Picture,errors='coerce')
NomineesDF_1970=Nominees[7].iloc[cotas(7,1970),:]

NomineesList=[NomineesDFs_1901_1909,NomineesDFs_1910_1969,NomineesDF_1970]
NomineesAllDFs=pd.concat(NomineesList)
NomineesAllDFs

Picture                          Name  \
1       NaN  Jacobus Henricus van 't Hoff   
2       NaN                  Emil Fischer   
3       NaN              Svante Arrhenius   
4       NaN                 Henri Moissan   
5       NaN            Zdenko Hans Skraup   
..      ...                           ...   
26      NaN    Charles Gilbert Overberger   
27      NaN   Henry Montague (Monty) Frey   
28      NaN           Jack Sylvester Hine   
29      NaN               Wolfgang Kirmse   
30      NaN                  Tetsuo Nozoe   

                                                 Born  \
1              August 30, 1852 Rotterdam, Netherlands   
2                 October 9, 1852 Euskirchen, Germany   
3                   February 19, 1859 Uppsala, Sweden   
4                    September 28, 1852 Paris, France   
5                March 3, 1850 Prague, Czech Republic   
..                                                ...   
26  October 12, 1920 Barnesboro, Pennsylvania, Uni...   
27                  February 27, 1929 London, England   
28  July 2, 1923  Coronado, California, United States   
29          June 26, 1930  Frankfurt am Main, Germany   
30                        May 16, 1902  Sendai, Japan   

                                           Died  \
1                 March 1, 1911 Berlin, Germany   
2                 July 15, 1919 Berlin, Germany   
3             October 2, 1927 Stockholm, Sweden   
4               February 20, 1907 Paris, France   
5            September 10, 1910 Vienna, Austria   
..                                          ...   
26     March 17, 1997  Ann Arbor, United States   
27                                          NaN   
28  July 6, 1988  Columbus, Ohio, United States   
29                                          NaN   
30                  April 4, 1996  Tokyo, Japan   

                       Years Nominated  \
1                                 1901   
2               1901, 1902, 1916, 1919   
3                     1901, 1902, 1903   
4   1901, 1902, 1903, 1904, 1905, 1906   
5                                 1901   
..                                 ...   
26                                1970   
27                                1970   
28                                1970   
29                                1970   
30                     1970, 1974[747]   

                                                Notes Unnamed: 6  
1           Won the 1901 Nobel Prize in Chemistry.[4]        NaN  
2   Won the 1902 Nobel Prize in Chemistry,[5] late...        NaN  
3   Won the 1903 Nobel Prize in Chemistry and nomi...        NaN  
4           Won the 1906 Nobel Prize in Chemistry.[8]        NaN  
5   Nominated the only time by Eduard Lippmann (18...        NaN  
..                                                ...        ...  
26  Jointly nominated with K.J.Freudenberg and Al....        NaN  
27  Jointly nominated the only time by Youssef Isk...        NaN  
28  Jointly nominated the only time by Youssef Isk...        NaN  
29  Jointly nominated the only time by Youssef Isk...        NaN  
30                                              [748]        NaN  

[640 rows x 7 columns]

### b.2) Limpieza del DataFrame resultante de la concatenación

In [143]:
NomineesAllDFs=pd.concat(NomineesList)
NomineesAllDFs=NomineesAllDFs.drop(columns=['Picture','Notes','Unnamed: 6'])
NomineesAllDFs=NomineesAllDFs.rename(columns={'Name':'Nominee','Born':'Birth_year','Died':'Death_year','Years Nominated':'Years_Nominated'})

NomineesAllDFs

Nominee  \
1   Jacobus Henricus van 't Hoff   
2                   Emil Fischer   
3               Svante Arrhenius   
4                  Henri Moissan   
5             Zdenko Hans Skraup   
..                           ...   
26    Charles Gilbert Overberger   
27   Henry Montague (Monty) Frey   
28           Jack Sylvester Hine   
29               Wolfgang Kirmse   
30                  Tetsuo Nozoe   

                                           Birth_year  \
1              August 30, 1852 Rotterdam, Netherlands   
2                 October 9, 1852 Euskirchen, Germany   
3                   February 19, 1859 Uppsala, Sweden   
4                    September 28, 1852 Paris, France   
5                March 3, 1850 Prague, Czech Republic   
..                                                ...   
26  October 12, 1920 Barnesboro, Pennsylvania, Uni...   
27                  February 27, 1929 London, England   
28  July 2, 1923  Coronado, California, United States   
29          June 26, 1930  Frankfurt am Main, Germany   
30                        May 16, 1902  Sendai, Japan   

                                     Death_year  \
1                 March 1, 1911 Berlin, Germany   
2                 July 15, 1919 Berlin, Germany   
3             October 2, 1927 Stockholm, Sweden   
4               February 20, 1907 Paris, France   
5            September 10, 1910 Vienna, Austria   
..                                          ...   
26     March 17, 1997  Ann Arbor, United States   
27                                          NaN   
28  July 6, 1988  Columbus, Ohio, United States   
29                                          NaN   
30                  April 4, 1996  Tokyo, Japan   

                       Years_Nominated  
1                                 1901  
2               1901, 1902, 1916, 1919  
3                     1901, 1902, 1903  
4   1901, 1902, 1903, 1904, 1905, 1906  
5                                 1901  
..                                 ...  
26                                1970  
27                                1970  
28                                1970  
29                                1970  
30                     1970, 1974[747]  

[640 rows x 4 columns]

In [144]:
NomineesAllDFs=pd.concat(NomineesList)
NomineesAllDFs=NomineesAllDFs.drop(columns=['Picture','Notes','Unnamed: 6'])
NomineesAllDFs=NomineesAllDFs.rename(columns={'Name':'Nominee','Born':'Birth_year','Died':'Death_year','Years Nominated':'Years_Nominated'})

# Manteniendo solo el año y el país en los valores de la columna Birth_year 
a=NomineesAllDFs['Birth_year'].str.split(',',expand=True)
a[1]=a[1].str.replace(pat=r'\D|\s',repl="",regex=True) #birth year
a[2]=a[2].str.replace(pat=r'\d|\s|\W',repl="",regex=True) #country
NomineesAllDFs['Birth_year']=a[1]
NomineesAllDFs["Country"]=a[2]

# Manteniendo solo el año en los valores de la columna Death_year
b=NomineesAllDFs['Death_year'].str.split(',',expand=True)
b[1]=b[1].str.replace(pat=r'\D|\s',repl="",regex=True) #death year
NomineesAllDFs['Death_year']=b[1]

#Resultado:
NomineesAllDFs

Nominee Birth_year Death_year  \
1   Jacobus Henricus van 't Hoff       1852       1911   
2                   Emil Fischer       1852       1919   
3               Svante Arrhenius       1859       1927   
4                  Henri Moissan       1852       1907   
5             Zdenko Hans Skraup       1850       1910   
..                           ...        ...        ...   
26    Charles Gilbert Overberger       1920       1997   
27   Henry Montague (Monty) Frey       1929        NaN   
28           Jack Sylvester Hine       1923       1988   
29               Wolfgang Kirmse       1930        NaN   
30                  Tetsuo Nozoe       1902       1996   

                       Years_Nominated        Country  
1                                 1901    Netherlands  
2               1901, 1902, 1916, 1919        Germany  
3                     1901, 1902, 1903         Sweden  
4   1901, 1902, 1903, 1904, 1905, 1906         France  
5                                 1901  CzechRepublic  
..                                 ...            ...  
26                                1970   Pennsylvania  
27                                1970        England  
28                                1970     California  
29                                1970        Germany  
30                     1970, 1974[747]          Japan  

[640 rows x 5 columns]

In [145]:
#Limpieza de la columna años de nominación
NomineesAllDFs['Years_Nominated']=NomineesAllDFs['Years_Nominated'].str.replace(pat=r'\s',repl="",regex=True)
NomineesAllDFs['Years_Nominated']=NomineesAllDFs['Years_Nominated'].str.split("[",expand=True)[0]
NomineesAllDFs

Nominee Birth_year Death_year  \
1   Jacobus Henricus van 't Hoff       1852       1911   
2                   Emil Fischer       1852       1919   
3               Svante Arrhenius       1859       1927   
4                  Henri Moissan       1852       1907   
5             Zdenko Hans Skraup       1850       1910   
..                           ...        ...        ...   
26    Charles Gilbert Overberger       1920       1997   
27   Henry Montague (Monty) Frey       1929        NaN   
28           Jack Sylvester Hine       1923       1988   
29               Wolfgang Kirmse       1930        NaN   
30                  Tetsuo Nozoe       1902       1996   

                  Years_Nominated        Country  
1                            1901    Netherlands  
2             1901,1902,1916,1919        Germany  
3                  1901,1902,1903         Sweden  
4   1901,1902,1903,1904,1905,1906         France  
5                            1901  CzechRepublic  
..                            ...            ...  
26                           1970   Pennsylvania  
27                           1970        England  
28                           1970     California  
29                           1970        Germany  
30                      1970,1974          Japan  

[640 rows x 5 columns]

In [146]:
#Ordenando el DF por año (una fila para cada año de nominación)
NomineesAllDFs_Long=NomineesAllDFs.copy()
NomineesAllDFs_Long['Years_Nominated']=NomineesAllDFs_Long['Years_Nominated'].str.split(",")
NomineesAllDFs_Long=NomineesAllDFs_Long.explode('Years_Nominated')
NomineesAllDFs_Long.reset_index(drop=True,inplace=True)
NomineesAllDFs_Long

Nominee Birth_year Death_year Years_Nominated  \
0     Jacobus Henricus van 't Hoff       1852       1911            1901   
1                     Emil Fischer       1852       1919            1901   
2                     Emil Fischer       1852       1919            1902   
3                     Emil Fischer       1852       1919            1916   
4                     Emil Fischer       1852       1919            1919   
...                            ...        ...        ...             ...   
2028   Henry Montague (Monty) Frey       1929        NaN            1970   
2029           Jack Sylvester Hine       1923       1988            1970   
2030               Wolfgang Kirmse       1930        NaN            1970   
2031                  Tetsuo Nozoe       1902       1996            1970   
2032                  Tetsuo Nozoe       1902       1996            1974   

          Country  
0     Netherlands  
1         Germany  
2         Germany  
3         Germany  
4         Germany  
...           ...  
2028      England  
2029   California  
2030      Germany  
2031        Japan  
2032        Japan  

[2033 rows x 5 columns]

In [147]:
#Transformación de la tabla a horizontal
NomineesAllDFs_Wide=pd.pivot_table(NomineesAllDFs_Long,
               values=["Nominee"], # values to use
               index=['Country'], # unit of analysis
               columns=['Years_Nominated'], # the values for NEW column
                aggfunc=lambda x: ' '.join(str(v) for v in x))

NomineesAllDFs_Wide

Nominee                                               \
Years_Nominated           152  153 1901 1902 1903 1904 1905 1906 1907 1908   
Country                                                                      
Argentina                 NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
Armenia                   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
Australia                 NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
Austria                   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
Belarus                   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
...                       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
WestRidingofYorkshire     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
Wiltshire                 NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
Wisconsin                 NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
nowPoland                 NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
nowUkraine                NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

                       ...                                        \
Years_Nominated        ...                                  1962   
Country                ...                                         
Argentina              ...                                   NaN   
Armenia                ...                                   NaN   
Australia              ...                                   NaN   
Austria                ...  Max Perutz Fritz Feigl Carl Djerassi   
Belarus                ...                                   NaN   
...                    ...                                   ...   
WestRidingofYorkshire  ...                                   NaN   
Wiltshire              ...                                   NaN   
Wisconsin              ...                                   NaN   
nowPoland              ...                                   NaN   
nowUkraine             ...                                   NaN   

                                                              \
Years_Nominated                          1963           1964   
Country                                                        
Argentina                                 NaN            NaN   
Armenia                                   NaN            NaN   
Australia                                 NaN            NaN   
Austria                Max Perutz Fritz Feigl  Carl Djerassi   
Belarus                                   NaN            NaN   
...                                       ...            ...   
WestRidingofYorkshire                     NaN            NaN   
Wiltshire                                 NaN            NaN   
Wisconsin                                 NaN            NaN   
nowPoland                        Fritz London            NaN   
nowUkraine                                NaN            NaN   

                                                                      \
Years_Nominated                                                 1965   
Country                                                                
Argentina                                                        NaN   
Armenia                                            Oganes K. Davtyan   
Australia                                  John Warcup Cornforth Jr.   
Austria                Herman Francis Mark Carl Djerassi Otto Kratky   
Belarus                                                          NaN   
...                                                              ...   
WestRidingofYorkshire                                            NaN   
Wiltshire                                                        NaN   
Wisconsin                                      James Elliot Carnahan   
nowPoland                                                        NaN   
nowUkraine                                            Erwin Chargaff   

                                                  \
Years_Nominated   

In [15]:
#Se observa que se presentan los años 152, 153 y 1974* que deben ser corregidos
#*Nota: la lista oficial de candidatos al nobel aparece 50 años luego de la fecha de nominación, por lo que aún no se han pubiclado los nominados en 1974

In [148]:
NomineesAllDFs_Long.loc[NomineesAllDFs_Long['Years_Nominated']=='152'] #Jaroslav Heyrovský
NomineesAllDFs_Long.loc[NomineesAllDFs_Long['Years_Nominated']=='153'] #Jaroslav Heyrovský
NomineesAllDFs_Long.loc[NomineesAllDFs_Long['Years_Nominated']=='1974'] #Tetsuo Nozoe

#Según la página de los premios Nobel, Jaroslav Heyrovský fue nominado en 1934,1938,1940,1944,1947,1950,1952,1953,1954,1955,1956,1957,1958,1959
    #link: https://www.nobelprize.org/prizes/chemistry/1959/heyrovsky/nominations/
    #Por tanto, fue un error de tipeo en Wikipedia (donde obtuve los datos)
    # Corrección de los datos
NomineesAllDFs.loc[NomineesAllDFs['Nominee']=='Jaroslav Heyrovský'] #row: 27
NomineesAllDFs.loc[27,'Years_Nominated']='1934,1938,1940,1944,1947,1950,1952,1953,1954,1955,1956,1957,1958,1959'

#Del mismo modo, solo se reporta que Tetsuo Nozoe fue nominado en 1970 y no en 1974
    #link: https://www.nobelprize.org/nomination/archive/show_people.php?id=15680
    # Corrección de los datos
NomineesAllDFs.loc[NomineesAllDFs['Nominee']=='Tetsuo Nozoe'] #row: 30
NomineesAllDFs.loc[30,'Years_Nominated']='1970'

#Resultado:
#Nueva transformación de la tabla a horizontal
NomineesAllDFs_Long=NomineesAllDFs.copy()
NomineesAllDFs_Long['Years_Nominated']=NomineesAllDFs_Long['Years_Nominated'].str.split(",")
NomineesAllDFs_Long=NomineesAllDFs_Long.explode('Years_Nominated')
NomineesAllDFs_Long.reset_index(drop=True,inplace=True)

NomineesAllDFs_Wide=pd.pivot_table(NomineesAllDFs_Long,
               values=["Nominee"], # values to use
               index=['Country'], # unit of analysis
               columns=['Years_Nominated'], # the values for NEW column
                aggfunc=lambda x: ' '.join(str(v) for v in x))

NomineesAllDFs_Wide

Nominee                                     \
Years_Nominated          1901 1902 1903 1904 1905 1906 1907 1908   
Country                                                            
Argentina                 NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
Armenia                   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
Australia                 NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
Austria                   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
Belarus                   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
...                       ...  ...  ...  ...  ...  ...  ...  ...   
WestRidingofYorkshire     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
Wiltshire                 NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
Wisconsin                 NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
nowPoland                 NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
nowUkraine                NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

                                                       ...              \
Years_Nominated                             1909 1910  ...        1961   
Country                                                ...               
Argentina                                    NaN  NaN  ...         NaN   
Armenia                                      NaN  NaN  ...         NaN   
Australia                                    NaN  NaN  ...         NaN   
Austria                Edmund Oscar von Lippmann  NaN  ...  Max Perutz   
Belarus                                      NaN  NaN  ...         NaN   
...                                          ...  ...  ...         ...   
WestRidingofYorkshire                        NaN  NaN  ...         NaN   
Wiltshire                                    NaN  NaN  ...         NaN   
Wisconsin                                    NaN  NaN  ...         NaN   
nowPoland                                    NaN  NaN  ...         NaN   
nowUkraine                                   NaN  NaN  ...         NaN   

                                                             \
Years_Nominated                                        1962   
Country                                                       
Argentina                                               NaN   
Armenia                                                 NaN   
Australia                                               NaN   
Austria                Max Perutz Fritz Feigl Carl Djerassi   
Belarus                                                 NaN   
...                                                     ...   
WestRidingofYorkshire                                   NaN   
Wiltshire                                               NaN   
Wisconsin                                               NaN   
nowPoland                                               NaN   
nowUkraine                                              NaN   

                                                              \
Years_Nominated                          1963           1964   
Country                                                        
Argentina                                 NaN            NaN   
Armenia                                   NaN            NaN   
Australia                                 NaN            NaN   
Austria                Max Perutz Fritz Feigl  Carl Djerassi   
Belarus                                   NaN            NaN   
...                                       ...            ...   
WestRidingofYorkshire                     NaN            NaN   
Wiltshire                                 NaN            NaN   
Wisconsin                                 NaN            NaN   
nowPoland                        Fritz London            NaN   
nowUkraine                                NaN            NaN   

                                                                      \
Years_Nominated                                                 1965   
Country                                                                
Argentina                           

In [149]:
#DF_Final
NomineesAllDFs_Long

Nominee Birth_year Death_year Years_Nominated  \
0     Jacobus Henricus van 't Hoff       1852       1911            1901   
1                     Emil Fischer       1852       1919            1901   
2                     Emil Fischer       1852       1919            1902   
3                     Emil Fischer       1852       1919            1916   
4                     Emil Fischer       1852       1919            1919   
...                            ...        ...        ...             ...   
2034   Henry Montague (Monty) Frey       1929        NaN            1958   
2035   Henry Montague (Monty) Frey       1929        NaN            1959   
2036           Jack Sylvester Hine       1923       1988            1970   
2037               Wolfgang Kirmse       1930        NaN            1970   
2038                  Tetsuo Nozoe       1902       1996            1970   

          Country  
0     Netherlands  
1         Germany  
2         Germany  
3         Germany  
4         Germany  
...           ...  
2034      England  
2035      England  
2036   California  
2037      Germany  
2038        Japan  

[2039 rows x 5 columns]

In [150]:
# Para realizar merge con las otras dos fuentes, se agruparán los nominados por año
Nominees_DF_2=NomineesAllDFs_Long.drop(columns=['Birth_year','Death_year'])
Nominees_DF_2=NomineesAllDFs_Long.groupby(['Years_Nominated'],as_index=False).agg({'Country':'first','Nominee':lambda x : ' , '.join(x.astype(str))})
Nominees_DF_2

Years_Nominated      Country  \
0             1901  Netherlands   
1             1902      Germany   
2             1903       Sweden   
3             1904       France   
4             1905       France   
..             ...          ...   
65            1966      England   
66            1967      England   
67            1968      England   
68            1969      England   
69            1970       France   

                                              Nominee  
0   Jacobus Henricus van 't Hoff , Emil Fischer , ...  
1   Emil Fischer , Svante Arrhenius , Henri Moissa...  
2   Svante Arrhenius , Henri Moissan , Marcellin B...  
3   Henri Moissan , Marcellin Berthelot , William ...  
4   Henri Moissan , Adolf von Baeyer , Theodore Wi...  
..                                                ...  
65  Christopher Kelk Ingold , Alexander Frumkin , ...  
66  Christopher Kelk Ingold , Alexander Frumkin , ...  
67  Christopher Kelk Ingold , Henry Eyring , Willi...  
68  Christopher Kelk Ingold , Walter Heitler , Hen...  
69  Paul Sabatier , Paul Walden , John Jacob Abel ...  

[70 rows x 3 columns]

In [152]:
#Limpieza final del DF
Nominees_DF_2['Year_of_award']=Nominees_DF_2['Years_Nominated']
Nominees_DF_2['Nominees']=Nominees_DF_2['Nominee']
Nominees_DF_2=Nominees_DF_2.drop(columns=['Years_Nominated','Nominee','Country'])
Nominees_DF_2

Year_of_award                                           Nominees
0           1901  Jacobus Henricus van 't Hoff , Emil Fischer , ...
1           1902  Emil Fischer , Svante Arrhenius , Henri Moissa...
2           1903  Svante Arrhenius , Henri Moissan , Marcellin B...
3           1904  Henri Moissan , Marcellin Berthelot , William ...
4           1905  Henri Moissan , Adolf von Baeyer , Theodore Wi...
..           ...                                                ...
65          1966  Christopher Kelk Ingold , Alexander Frumkin , ...
66          1967  Christopher Kelk Ingold , Alexander Frumkin , ...
67          1968  Christopher Kelk Ingold , Henry Eyring , Willi...
68          1969  Christopher Kelk Ingold , Walter Heitler , Hen...
69          1970  Paul Sabatier , Paul Walden , John Jacob Abel ...

[70 rows x 2 columns]

## c) Fuente 3

In [17]:
import pandas as pd

AffiliationLink="https://en.wikipedia.org/wiki/List_of_Nobel_laureates_by_university_affiliation"
Affiliation=pd.read_html(AffiliationLink,header=0,flavor='bs4')
type(Affiliation)

list

In [40]:
Affiliation[0]

Laureate              Discipline  Year of award  \
0           Alexei Abrikosov                 Physics           2003   
1               Edgar Adrian  Physiology or Medicine           1932   
2            Pierre Agostini                 Physics           2023   
3                 Peter Agre               Chemistry           2003   
4              Isamu Akasaki                 Physics           2014   
..                       ...                     ...            ...   
806            Frits Zernike                 Physics           1953   
807             Ahmed Zewail               Chemistry           1999   
808             Karl Ziegler               Chemistry           1963   
809      Rolf M. Zinkernagel  Physiology or Medicine           1996   
810  Richard Adolf Zsigmondy               Chemistry           1925   

                                Affiliation  
0               Argonne National Laboratory  
1                   University of Cambridge  
2                     Ohio State University  
3          Johns Hopkins School of Medicine  
4                          Meijo University  
..                                      ...  
806                 University of Groningen  
807      California Institute of Technology  
808  Max Planck Institute for Coal Research  
809                    University of Zurich  
810                 University of Göttingen  

[811 rows x 4 columns]

In [41]:
Affiliation[0].columns

Index(['Laureate', 'Discipline', 'Year of award', 'Affiliation'], dtype='object')

In [159]:
#Creando un DF con solo los ganadores al premio Nobel de Química
Affiliation_DF=Affiliation[0].loc[Affiliation[0]['Discipline']=='Chemistry']

#Eliminado la columna 'Discipline' del nuevo DF
Affiliation_DF.drop(columns=['Discipline'],inplace=True)

# Renombrando la columna "Year of award"
Affiliation_DF = Affiliation_DF.rename(columns={"Year of award":"Year_of_award"})
Affiliation_DF

C:\Users\Patiy Li Yang\AppData\Local\Temp\ipykernel_11844\356506548.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Affiliation_DF.drop(columns=['Discipline'],inplace=True)


Laureate  Year_of_award  \
3                 Peter Agre           2003   
7                 Kurt Alder           1950   
14             Sidney Altman           1989   
19     Christian B. Anfinsen           1972   
23            Frances Arnold           2018   
..                       ...            ...   
798            Akira Yoshino           2019   
799            Akira Yoshino           2019   
807             Ahmed Zewail           1999   
808             Karl Ziegler           1963   
810  Richard Adolf Zsigmondy           1925   

                                Affiliation  
3          Johns Hopkins School of Medicine  
7                     University of Cologne  
14                          Yale University  
19            National Institutes of Health  
23       California Institute of Technology  
..                                      ...  
798                             Asahi Kasei  
799                        Meijo University  
807      California Institute of Technology  
808  Max Planck Institute for Coal Research  
810                 University of Göttingen  

[211 rows x 3 columns]

In [43]:
#Ordenando los datos de forma ascendente por año
Affiliation_DF.sort_values(by=['Year_Won'], inplace=True)
Affiliation_DF.reset_index(inplace=True,drop=True)
Affiliation_DF

Laureate  Year_Won  \
0    Jacobus Henricus van 't Hoff      1901   
1                    Emil Fischer      1902   
2                Svante Arrhenius      1903   
3                  William Ramsay      1904   
4                Adolf von Baeyer      1905   
..                            ...       ...   
206              Carolyn Bertozzi      2022   
207              Morten P. Meldal      2022   
208          Karl Barry Sharpless      2022   
209                 Louis E. Brus      2023   
210                Moungi Bawendi      2023   

                                Affiliation  
0             Humboldt University of Berlin  
1             Humboldt University of Berlin  
2                      Stockholm University  
3                 University College London  
4    Ludwig Maximilian University of Munich  
..                                      ...  
206         Howard Hughes Medical Institute  
207                University of Copenhagen  
208                        Scripps Research  
209                     Columbia University  
210   Massachusetts Institute of Technology  

[211 rows x 3 columns]

## d) Unión (merging) de los tres DFs

In [168]:
Nominees_DF_2.merge(Laureates_DF,how='right',left_on='Year_of_award	',right_on='Year_of_award')

KeyError: 'Year_of_award\t'

In [162]:
Nominees_DF_2.merge(Laureates_DF,how='right',on='Year_of_award')

Year_of_award                                           Nominees
0           1901  Jacobus Henricus van 't Hoff , Emil Fischer , ...
1           1902  Emil Fischer , Svante Arrhenius , Henri Moissa...
2           1903  Svante Arrhenius , Henri Moissan , Marcellin B...
3           1904  Henri Moissan , Marcellin Berthelot , William ...
4           1905  Henri Moissan , Adolf von Baeyer , Theodore Wi...
..           ...                                                ...
65          1966  Christopher Kelk Ingold , Alexander Frumkin , ...
66          1967  Christopher Kelk Ingold , Alexander Frumkin , ...
67          1968  Christopher Kelk Ingold , Henry Eyring , Willi...
68          1969  Christopher Kelk Ingold , Walter Heitler , Hen...
69          1970  Paul Sabatier , Paul Walden , John Jacob Abel ...

[70 rows x 2 columns]

In [164]:
Laureates_DF.head(70)

Year_of_award                       Laureate         Country  \
0            1901  Jacobus Henricus van 't Hoff      Netherlands   
1            1902          Hermann Emil Fischer          Germany   
2            1903       Svante August Arrhenius           Sweden   
3            1904            Sir William Ramsay   United Kingdom   
4            1905              Adolf von Baeyer          Germany   
..            ...                            ...             ...   
65           1962          Max Ferdinand Perutz   United Kingdom   
66           1962          John Cowdery Kendrew   United Kingdom   
67           1963                  Karl Ziegler     West Germany   
68           1963                  Giulio Natta            Italy   
69           1964      Dorothy Crowfoot Hodgkin   United Kingdom   

                                            Rationale Birth_year Death_year  
0   "[for his] discovery of the laws of chemical d...       1852       1911  
1      "[for] his work on sugar and purine syntheses"       1852       1919  
2     "[for] his electrolytic theory of dissociation"       1859       1927  
3   "[for his] discovery of the inert gaseous elem...       1852       1916  
4   "[for] the advancement of organic chemistry an...       1835       1917  
..                                                ...        ...        ...  
65  "for their studies of the structures of globul...       1914       2002  
66  "for their studies of the structures of globul...       1917       1997  
67  "for their discoveries in the field of the che...       1898       1973  
68  "for their discoveries in the field of the che...       1903       1979  
69  "for her determinations by X-ray techniques of...       1910       1994  

[70 rows x 6 columns]

## d.1) Unión de Laureates_DF y Affiliation_DF

**Problema 1**: No es posible realizar un 'merge' por año o ganador, pues Affiliation_DF contiene más filas que Laureates_DF (211 vs. 202). Revisando la [fuente 3](https://en.wikipedia.org/wiki/List_of_Nobel_laureates_by_university_affiliation), ciertos ganadores presentan más de una afiliación y se ha creado una extra fila para cada afiliación. Por tanto, se deberán borrar los duplicados en Affiliation_DF.

In [44]:
Laureates_DF.info(), Affiliation_DF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202 entries, 0 to 201
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Year_Won    202 non-null    int64 
 1   Laureate    202 non-null    object
 2   Country     202 non-null    object
 3   Rationale   202 non-null    object
 4   Birth_year  194 non-null    object
 5   Death_year  202 non-null    object
dtypes: int64(1), object(5)
memory usage: 9.6+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211 entries, 0 to 210
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Laureate     211 non-null    object
 1   Year_Won     211 non-null    int64 
 2   Affiliation  211 non-null    object
dtypes: int64(1), object(2)
memory usage: 5.1+ KB


(None, None)

In [69]:
# Agrupación de las filas correspondienets al mismo ganador por año, agrupando las afiliaciones por comas en la tercera columna
Affiliation_DF=Affiliation_DF.groupby(['Laureate','Year_Won'],as_index=False).agg({'Year_Won':'first','Affiliation':lambda x : ','.join(x.astype(str))})

#Ordenando los datos de forma ascendente por año
Affiliation_DF.sort_values(by=['Year_Won'], inplace=True)
Affiliation_DF.reset_index(inplace=True,drop=True)
Affiliation_DF

Laureate  Year_Won  \
0    Jacobus Henricus van 't Hoff      1901   
1                    Emil Fischer      1902   
2                Svante Arrhenius      1903   
3                  William Ramsay      1904   
4                Adolf von Baeyer      1905   
..                            ...       ...   
185              Carolyn Bertozzi      2022   
186              Morten P. Meldal      2022   
187          Karl Barry Sharpless      2022   
188                 Louis E. Brus      2023   
189                Moungi Bawendi      2023   

                                           Affiliation  
0                        Humboldt University of Berlin  
1                        Humboldt University of Berlin  
2                                 Stockholm University  
3                            University College London  
4               Ludwig Maximilian University of Munich  
..                                                 ...  
185  Stanford University,Howard Hughes Medical Inst...  
186                           University of Copenhagen  
187                                   Scripps Research  
188                                Columbia University  
189              Massachusetts Institute of Technology  

[190 rows x 3 columns]

In [94]:
# Filas donde se observan duplicados en las columnas 'Laureate' y 'Year_Won'
Affiliation_DF[Affiliation_DF.duplicated(['Laureate','Year_Won'],keep=False)]

#Notar que se ha seleccionado ambas columnas, pues Frederick Sanger y 	Karl Barry Sharpless han ganado el premio Nobel en años distintos

Empty DataFrame
Columns: [Laureate, Year_Won, Affiliation]
Index: []

In [50]:
# Revisando nuevamente los duplicados: Se han logrado corregir satisfactoriamente 
Affiliation_DF[Affiliation_DF.duplicated(['Laureate','Year_Won'],keep=False)]

Empty DataFrame
Columns: [Laureate, Year_Won, Affiliation]
Index: []

**Problema 2**: Se han logrado corregir los duplicados en Affiliation_DF. Sin embargo, el número de filas se ha reducido a 190, cuando Laureates_DF contiene 202.

In [51]:
# Revisando duplicados en Laureates_DF
Laureates_DF[Laureates_DF.duplicated(['Laureate'],keep=False)]

Year_Won             Laureate         Country  \
16       1916          Not awarded     Not awarded   
17       1917          Not awarded     Not awarded   
19       1919          Not awarded     Not awarded   
24       1924          Not awarded     Not awarded   
35       1933          Not awarded     Not awarded   
45       1940          Not awarded     Not awarded   
46       1941          Not awarded     Not awarded   
47       1942          Not awarded     Not awarded   
69       1958    Frederick Sanger   United Kingdom   
103      1980    Frederick Sanger   United Kingdom   
145      2001  K. Barry Sharpless    United States   
198      2022  K. Barry Sharpless    United States   

                                             Rationale Birth_year Death_year  
16                                         Not awarded       None      Alive  
17                                         Not awarded       None      Alive  
19                                         Not awarded       None      Alive  
24                                         Not awarded       None      Alive  
35                                         Not awarded       None      Alive  
45                                         Not awarded       None      Alive  
46                                         Not awarded       None      Alive  
47                                         Not awarded       None      Alive  
69   "for his work on the structure of proteins, es...       1918       2013  
103  "for their contributions concerning the determ...       1918       2013  
145  "for his work on chirally catalysed oxidation ...       1941      Alive  
198  "for the development of click chemistry and bi...       1941      Alive

In [160]:
# Hay 8 años, donde no se otorgó el premio. Se eliminarán dichas filas
# Con esto, el número de filas se ha reducido a 194
Laureates_DF.drop(Laureates_DF[Laureates_DF['Laureate'] == 'Not awarded'].index, inplace = True)
Laureates_DF.reset_index(inplace=True,drop=True)
Laureates_DF

Year_of_award                       Laureate         Country  \
0             1901  Jacobus Henricus van 't Hoff      Netherlands   
1             1902          Hermann Emil Fischer          Germany   
2             1903       Svante August Arrhenius           Sweden   
3             1904            Sir William Ramsay   United Kingdom   
4             1905              Adolf von Baeyer          Germany   
..             ...                            ...             ...   
189           2022                 Morten Meldal          Denmark   
190           2022            K. Barry Sharpless    United States   
191           2023             Moungi G. Bawendi    United States   
192           2023                 Louis E. Brus    United States   
193           2023                 Alexey Ekimov           Russia   

                                             Rationale Birth_year Death_year  
0    "[for his] discovery of the laws of chemical d...       1852       1911  
1       "[for] his work on sugar and purine syntheses"       1852       1919  
2      "[for] his electrolytic theory of dissociation"       1859       1927  
3    "[for his] discovery of the inert gaseous elem...       1852       1916  
4    "[for] the advancement of organic chemistry an...       1835       1917  
..                                                 ...        ...        ...  
189  "for the development of click chemistry and bi...       1954      Alive  
190  "for the development of click chemistry and bi...       1941      Alive  
191  "for the discovery and synthesis of quantum dots"       1961      Alive  
192  "for the discovery and synthesis of quantum dots"       1943      Alive  
193  "for the discovery and synthesis of quantum dots"       1945      Alive  

[194 rows x 6 columns]

In [108]:
# Identificar filas no comunes en ambos DF:
# Se deben encontrar 4 nombres no comunes que deberán ser incluidos manualmente en Affiliation_DF

# los nombres no se encuentran en el mismo formato, por lo que se deberá realizar un fuzzy merge
Laureates_set=set(Laureates_DF.Laureate)
Affiliation_set=set(Affiliation_DF.Laureate)

# Tomar un nombre en Affiliation_set y buscar el más similar en Laureates_set
from thefuzz import process as fz
[(f,fz.extractOne(f,Laureates_set)) for f in sorted(Affiliation_set)
   if fz.extractOne(f,Laureates_set)[1]>86]


{'Aaron Ciechanover ',
 'Aaron Klug ',
 'Ada E. Yonath ',
 'Adolf Friedrich Johann Butenandt ',
 'Adolf Otto Reinhold Windaus ',
 'Adolf von Baeyer ',
 'Ahmed Zewail ',
 'Akira Suzuki ',
 'Akira Yoshino ',
 'Alan G. MacDiarmid ',
 'Alan J. Heeger ',
 'Alexey Ekimov ',
 'Alfred Werner ',
 'Archer John Porter Martin ',
 'Arieh Warshel ',
 'Arne Wilhelm Kaurin Tiselius ',
 'Arthur Harden ',
 'Artturi Ilmari Virtanen ',
 'Avram Hershko ',
 'Aziz Sancar ',
 'Ben Feringa ',
 'Benjamin List ',
 'Brian Kobilka ',
 'Carl Bosch ',
 'Carolyn Bertozzi ',
 'Charles J. Pedersen ',
 'Christian B. Anfinsen ',
 'Dan Shechtman ',
 'David W.C. MacMillan ',
 'Derek H. R. Barton ',
 'Donald J. Cram ',
 'Dorothy Crowfoot Hodgkin ',
 'Dudley R. Herschbach ',
 'Eduard Buchner ',
 'Edwin Mattison McMillan ',
 'Ei-ichi Negishi ',
 'Elias James Corey ',
 'Emmanuelle Charpentier ',
 'Eric Betzig ',
 'Ernest Rutherford ',
 'Ernst Otto Fischer ',
 'Frances Arnold ',
 'Francis William Aston ',
 'Frank Sherwood Rowla